81816 clustered purge improvements

chguocloudant opened the pull request:
Update for Clustered Purge Improvement

 - add verify_index_exists function to check if the index exist.
 - update utc_timestamp function.
 - add some test cases.

BugzId: 81816

chguocloudant committed:
Update dreyfus for the pluggable storage engine API change

- use couch_db:fold_changes/5
- use load_docs/2

BugzID: 68280

chguocloudant committed:
Update dreyfus for clustered purge

- add 2 APIs(set_purge_seq/2, get_purge_seq/1) in dreyfus clouseau_rpc.
- write test logic to cover changed code.
- purge index and update purge seq in clouseau.

BugzId: 68280

chguocloudant committed:
Update for test cases

- add 2 test cases(index purging for docs with more then 1 revs).

BugzId: 68280

chguocloudant committed:
Adjust implement of index purge

- using delete or update to maintain index, avoid delete first and then update it.

BugzId: 68280

chguocloudant committed:
Avoid unnecessary operation to index

- add a List to store finished {Id, Rev}, avoid unnecessary index update or delete.
- add some test cases to test index update and delete.

BugzId: 68280

chguocloudant committed:
Update dreyfus for details

- extract a new function to delete or update index.
- remove useless function.
- take more idiomatic and faster way to update list.

BugzId: 68280

chguocloudant committed:
Update dreyfus for Clustered Purge Improvements

- create/update/delete local purge doc.
- add verify_index_exists function to check if the index exist.
- add a new api to get index root dir from clouseau.
- adjust cleanup code for search index.
- add some test cases.

BugzId: 81816

mayya-sharipova commented on src/dreyfus_fabric_cleanup.erl:
```diff
@@ -26,10 +26,20 @@ go(DbName) ->
     clouseau_rpc:cleanup(DbName, ActiveSigs),
+    ok.
+
+active_sigs(#doc{body={Fields}}=Doc) ->
```
Why `active_sigs` has been moved? Looks like a new code.

chguocloudant replied:
Hi Mayya. I didn't change this function. I think it is a mistake by compare tool.

mayya-sharipova commented on src/dreyfus_fabric_cleanup.erl:
```diff
@@ -26,10 +26,20 @@ go(DbName) ->
-    %clean local doc
-    BaseDir = config:get("couchdb", "search_index_dir", "/srv/search_index"),
+cleanup_local_doc(DbName, ActiveSigs) ->
+    {ok, BaseDir} = clouseau_rpc:get_root_dir(),
```
What is the reason that you are not using config to get `BaseDir`?

chguocloudant replied:
Hi Mayya. 
a) I checked the config file and there is no record of `search_index_dir`.
b) The search index dir is controlled by `clouseau` in fact, so I think it is better to get `BaseDir` from `clouseau`.

mayya-sharipova commented on src/dreyfus_fabric_cleanup.erl:
```diff
@@ -35,3 +37,41 @@ active_sigs(#doc{body={Fields}}=Doc) ->
          Index#index.sig
      end || IndexName <- IndexNames].
+
+cleanup_local_doc(DbName, ActiveSigs) ->
```
I am not sure if we need this complex code to delete _local/purge-search ... docs. Why do we need to open all  8 shards dbs and do it in a loop?
I would just use the following code:

```
cleanup_local_doc(DbName, ActiveSigs) ->
   Opts = [{start_key, list_to_binary(?LOCAL_DOC_PREFIX ++ "purge-search-")},
        {end_key_gt, list_to_binary(?LOCAL_DOC_PREFIX ++ "purge-search1")}],
       % use fabric all_docs with Ops to get  all _local/purge-search docs
       {ok, LocalPDocs} = fabric:all_docs(DbName,....)
        % filter those the need to be deleted
        % use fabric update_docs for docs that need to be removed
        fabric:update_docs(

``

chguocloudant replied:
Hi Mayya. The function `fabric:all_docs` cannot collect all the local docs because there is an error in function `fold_local_docs` which is used.

mayya-sharipova commented on src/dreyfus_fabric_cleanup.erl:
```diff
@@ -18,13 +18,15 @@
     {ok, DesignDocs} = fabric:design_docs(DbName),
     ActiveSigs = lists:usort(lists:flatmap(fun active_sigs/1,
         [couch_doc:from_json_obj(DD) || DD <- DesignDocs])),
+    cleanup_local_doc(DbName, ActiveSigs),
```
Would `cleanup_local_purge_docs` be a better name?

chguocloudant replied:
yes, you are right. I will modify it later. Thank you.

mayya-sharipova commented on src/dreyfus_fabric_cleanup.erl:
```diff
@@ -35,3 +37,28 @@ active_sigs(#doc{body={Fields}}=Doc) ->
+        #doc{id = DocId, deleted=true}
+    end, DeadDirs),
+    fabric:update_docs(DbName, DeleteDocs, [?ADMIN_CTX]),
+    ok.
```
LGTM (looks good to me)!

mayya-sharipova commented on src/dreyfus_index.erl:
```diff
@@ -118,6 +119,16 @@ init({DbName, Index}) ->
+    DocId = dreyfus_util:get_local_purge_doc_id(Index#index.sig),
+    case couch_db:open_doc(Db, DocId) of
+        {not_found, _} ->
+            DocContent = dreyfus_util:get_local_purge_doc_body(Db, DocId, 0, Index),
```
@chguocloudant  I think, Initial purge_seq that you pass to `dreyfus_util:get_local_purge_doc_body` should be NOT 0, but the database's purge_seq.

chguocloudant replied:
I think if the local doc is not exist the purge seq is 0.
But I will modify it and the code will be more strong.
Thank you.

mayya-sharipova commented on src/dreyfus_index_updater.erl:
```diff
@@ -64,10 +66,55 @@ update(IndexPid, Index) ->
+            {ok, Acc0}
+        end,
+        couch_db:fold_purged_docs(Db, IdxPurgeSeq, FoldFun, [], [])
+    after
```
@chguocloudant  A new format of `fold_purged_docs` function is following:
```
FoldFun = fun({PurgeSeq ,_UUID, Id, Revs}, Acc) ->  ... NewAcc end,
{ok, PurgedIdsRevs} = couch_db:fold_purged_docs(Db, IdxPurgeSeq, FoldFun, [], []),
```

chguocloudant replied:
Thank you Mayya, will modify it later.

mayya-sharipova commented on src/dreyfus_index_updater.erl:
```diff
@@ -64,10 +66,55 @@ update(IndexPid, Index) ->
+            Acc0 = case couch_db:get_full_doc_info(Db, Id) of
+                not_found ->
+                    ok = clouseau_rpc:delete(IndexPid, Id),
+                    update_local_doc(Db, Index, PurgeSeq),
```
@chguocloudant  Would a name `update_local_purge_doc` be better? 

Do you think we can do `update_local_doc` and `clouseau_rpc:set_purge_seq(IndexPid, PurgeSeq)` outside of the loop with the final purge_seq?

chguocloudant replied:
Hi @mayya-sharipova . I want to confirm that what is the order of the purge seq.
Should I check and compare to get the largest purge seq?
Or the first/last seq is the final(largest) purge seq?

mayya-sharipova commented on src/dreyfus_index_updater.erl:
```diff
@@ -64,10 +66,55 @@ update(IndexPid, Index) ->
+                FDI ->
+                    DI = couch_doc:to_doc_info(FDI),
+                    #doc_info{id=Id, revs=[#rev_info{rev=Rev}|_]} = DI,
+                    case lists:member({Id, Rev}, Acc) of
```
It is not clear what you are checking here  with this `case` statement - can you put a comment for this

chguocloudant replied:
If a doc has more Revs and purged twice or more, there will be twice or more purge seq for this doc. So the index will be update/delete twice or more. In fact it will be OK only once.
So I check the `{Id, Rev}` to keep update/delete once.
This change is in another PR(#11).
The new PR(#12) I created is based on it and the commit history is kept.

mayya-sharipova replied:
@chguocloudant I still think this check is not necessary. First, a user wants to delete several revs of a particular doc, the user should provided them: _purge  ID:[Rev1, Rev2, Rev3]; and this will create only one record of `{PurgeSeq, _UUID, Id, _Revs}` in purge trees.
Second, even if there are several requests with the same ID, nothing will happens if you call `clouseau_rpc:delete(IndexPid, Id)` with the same `Id` several times (I think).

chguocloudant replied:
Hi @mayya-sharipova .
The `Acc` is a list of {Id, Rev}. And the `Rev` is not the purged `Rev`, it is the final `Rev`.
For example. A doc(DocA) has Rev1, Rev2, Rev3, and purged Rev1 and Rev2. Then {DocA, Rev3} will be kept in `Acc`.
And if we call `clouseau_rpc:delete` twice with same Id, I am not sure what will happen in `Lucene` but `IndexWriter.deleteDocuments` is called twice directly.

mayya-sharipova commented on src/dreyfus_index_updater.erl:
```diff
@@ -64,10 +66,55 @@ update(IndexPid, Index) ->
+                    case lists:member({Id, Rev}, Acc) of
+                        true -> Acc;
+                        false ->
+                            update_or_delete_index(IndexPid, Db, DI, Proc),
```
would instead of `update_or_delete_index` a better name be `update_or_delete_doc`?

chguocloudant replied:
I think `update_or_delete_index` will be OK. It might be ambiguous if it is named as a doc.

mayya-sharipova commented on src/dreyfus_index_updater.erl:
```diff
@@ -31,6 +31,7 @@ update(IndexPid, Index) ->
     erlang:put(io_priority, {view_update, DbName, IndexName}),
     {ok, Db} = couch_db:open_int(DbName, []),
     try
+        {ok, ExcludeIdRevs} = purge_index(Db, IndexPid, Index),
```
@chguocloudant What is the reason you are collecting this `ExcludeIdRevs`? Would a simplier approach be just do `clouseau_rpc:delete(IndexPid, Id);` for all docs in `purge_index` function, and then keep `load_docs` function as it was before without any check `case lists:member({Id, Rev}, ExcludeIdRevs) `.

With my suggestion, If a doc is fully purged, it will be deleted in `purge_index` function using `clouseau_rpc:delete`. If a doc is partially purged, we will first also delete it in `purge_index`, and then the doc's new revision should come through changes to the `load_docs` function, and should be update there.

chguocloudant replied:
Hi @mayya-sharipova . If one doc is purged, there will be a purge seq and db update seq. The index delete/update will be run twice. So we collect the `ExcludeIdRevs` in `purge_index` to avoid this situation.
This change is in another PR(<https://github.com/cloudant-labs/dreyfus/pull/11>). 
The new PR(#12) I created is based on it and the commit history is kept.

mayya-sharipova replied:
@chguocloudant Yes, exactly. If a doc is partially purged, if you will get both a purge_seq change (where you can always just a delete doc from clouseau), and update_seq change where you can insert a new doc to clouseau.  If during purge_index, you only delete docs, the delete/update part will not run twice

chguocloudant replied:
Hi @mayya-sharipova .
1. I used to delete the index first and update(rebuild) it in PR(#11). But Robert Newson asked me to `go from its current state to the new state without an intermediate state`, and `Here we will make the intermediate state, where the Documents in the Lucene index are removed, visible to search results.` So we delete or delete index instead of delete first and then update.
2. Based 1, we found that the index maybe delete/update twice, first in `purge_index` and second in `load_docs`. And update and delete in `clouseau` is expensive so we add an `ExcludeList` to avoid the unnecessary operations.

mayya-sharipova replied:
@chguocloudant Thanks for explanation, your current code makes more sense now.  Please keep it.

mayya-sharipova commented on src/dreyfus_index.erl:
```diff
@@ -119,11 +119,12 @@ init({DbName, Index}) ->
         {not_found, _} ->
-            DocContent = dreyfus_util:get_local_purge_doc_body(Db, DocId, 0, Index),
+            DocContent = dreyfus_util:get_local_purge_doc_body(Db, DocId, PurgeSeq, Index),
             couch_db:update_doc(Db, DocContent, []);
```
@chguocloudant We need to think here, if the  initial purge_seq should be the **database's** purge_seq (or from clouseau).  

Consider this, you are having `maybe_create_local_purge_doc` function for the  upgrade. Before this upgrade `clouseau` was not storing any purge_seq (that what makes this index different from hastings).  So, initially you would like to get purge_seq from a database. 

You can put line #124 after line #126, so to request purge_seq only when necessary.

mayya-sharipova commented on src/dreyfus_index.erl:
```diff
@@ -388,18 +389,18 @@ verify_index_exists(Options) ->
+                {ok, #index{sig=Sig0}} = dreyfus_index:design_doc_to_index(DDoc, IdxName),
+                Sig == Sig0
+            catch _:_ ->
+                couch_log:warning("Something wrong happened, please check it. "
```
I think instead of `couch_log:warning` returning false is enough.  Also, a log contains thousands of messages, and when writing smth to log, you should specify the content what is the problem, and where it is coming from; otherwise it would be very difficult to debut it.

chguocloudant replied:
Thank you Mayya. I will remove the log.

mayya-sharipova commented on src/dreyfus_index.erl:
```diff
@@ -118,6 +119,17 @@ init({DbName, Index}) ->
 
+maybe_create_local_purge_doc(Db, IndexPid, Index) ->
+    DocId = dreyfus_util:get_local_purge_doc_id(Index#index.sig),
+    {ok, PurgeSeq} = clouseau_rpc:get_purge_seq(IndexPid),
```
@chguocloudant We need to think here, if the initial purge_seq should be the database's purge_seq (or from clouseau).

Consider this, you are having maybe_create_local_purge_doc function for the upgrade. Before this upgrade clouseau was not storing any purge_seq (that what makes this index different from hastings). So, initially you would like to get purge_seq from a database.

Or if you have different plan for upgrade, please suggest it.

You can put line #124 after line #126, so to request purge_seq only when necessary.

chguocloudant replied:
Hi Mayya. 
clouseau will return the default value 0 if it has not yet store a purge_seq.
And if the local purge doc is not exist here, it means purge_index is not called. The purge_seq should be 0 here.
And I am not sure the means "get purge_seq from a database". The purge seq is not kept in database now.
And the line #124 should be move to line #126, thank you, I will move this line.

chguocloudant replied:
Understand now and committed it already.

mayya-sharipova commented on src/dreyfus_index_updater.erl:
```diff
@@ -101,14 +100,16 @@ purge_index(Db, IndexPid, Index) ->
             end,
-            clouseau_rpc:set_purge_seq(IndexPid, PurgeSeq),
-            {ok, Acc0}
+            [Acc0, PurgeSeq]
```
did you want to use a tuple here: `{Acc0, PurgeSeq}? I have not see a list to be used like this

chguocloudant replied:
Thank you Mayya. I will modify it.

davisp commented on src/dreyfus_fabric_cleanup.erl:
```diff
@@ -18,13 +18,15 @@
 -include("dreyfus.hrl").
 -include_lib("mem3/include/mem3.hrl").
 -include_lib("couch/include/couch_db.hrl").
+-include_lib("couch_mrview/include/couch_mrview.hrl").
```
What's this import for? This links dreyfus to couch_mrview which is not a good idea.

chguocloudant replied:
Thanks, it should be removed. I will modify it.

davisp commented on src/dreyfus_index.erl:
```diff
@@ -369,3 +383,23 @@ group2_int(Pid, QueryArgs0) ->
     clouseau_rpc:info(Pid).
+
+verify_index_exists(Options) ->
+    DbName = dreyfus_util:get_value_from_options(<<"db_name">>, Options),
```
These keys are different than what the change to couch_mrview added. We should settle on the naming scheme (even though this would work fine, consistency is always better).

chguocloudant replied:
I will keep them same to hastings and couch_mrview.

davisp commented on src/dreyfus_util.erl:
```diff
@@ -166,6 +167,46 @@ time(Metric, {M, F, A}) when is_list(Metric) ->
         couch_stats:update_histogram([dreyfus | Metric],  Length)
     end.
 
+utc_timestamp() ->
```
We should consider adding this function to couch_util since its used multiple times. We may also want to check if something like this already exists.

chguocloudant replied:
We didn't find any timestamp function in couch_util yet. I will modify it after it added in couch_util.

davisp commented on src/dreyfus_util.erl:
```diff
@@ -166,6 +167,46 @@ time(Metric, {M, F, A}) when is_list(Metric) ->
+    end.
+
+get_local_purge_doc_id(Sig) ->
+    <<?LOCAL_DOC_PREFIX, "purge-search-", Sig/binary>>.
```
Should we call this dreyfus to be a bit more specific than search?

chguocloudant replied:
Thank you. I will modify "search" to "dreyfus".

davisp commented on src/dreyfus_index_updater.erl:
```diff
@@ -77,15 +129,14 @@ load_docs(FDI, _, {I, IndexPid, Db, Proc, Total, LastCommitTime}=Acc) ->
             case Fields1 of
-                [] -> ok = clouseau_rpc:delete(IndexPid, Id);
-                _  -> ok = clouseau_rpc:update(IndexPid, Id, Fields1)
+                [] ->
```
Undo this hunk since its not changing anything except formatting.

chguocloudant replied:
Thank you. I will modify it.

davisp commented on src/dreyfus_index_updater.erl:
```diff
@@ -64,10 +68,58 @@ update(IndexPid, Index) ->
+    try
+        true = proc_prompt(Proc, [<<"add_fun">>, Index#index.def]),
+        FoldFun = fun({PurgeSeq, _UUID, Id, _Revs}, {Acc, _}) ->
+            Acc0 = case couch_db:get_full_doc_info(Db, Id) of
```
You shouldn't need to be loading the document info here at all. You just need to delete the index entry if it exists and not cause an error if it doesn't exist. That way your update code stays the same and your change will be considerably smaller and less prone to breakage. Adding data to the index while operating over the purge sequences isn't how this was designed so there may very well be consequences to adding data based on two orderings that haven't been thought of.

chguocloudant replied:
Hi @davisp . 
The last PR (#11) is reviewed by @rnewson . And we realized that  if we only delete index in purging, we will make a intermediate state where the Documents in the Lucene index are removed, visible to search results.

rnewson replied:
@davisp This is to prevent the index passing through an intermediate stage where the old rows are purged but the replacement rows are not yet inserted, a state that it should not be in, even temporarily. This wouldn't matter if clustered purge was just for the very rare genuine cases of purge, but we intend to use it a lot, and as part of 'hard delete'.

davisp reviewed (changes requested):
This is mostly good. I think the operations during purge are wrong though. We've never added entries to an index during purge and that's not how its designed so doing so worries me quite a bit. I can't think of how it would fail but I can't convince myself there's not a bug there. However if we go the normal route of just removing any entries there then we don't have to worry about it.

chguocloudant committed:
Address the review comments

- check the sig format
- make the field name in local purge doc same to geo and mrview
- use timestamp function in couch_util
- adjust the test cases

BugzId: 81816
